internal/builder/builder.go

```diff
@@ -0,0 +1,89 @@
+package builder
+
+import (
+	"bytes"
+	gojson "encoding/json"
+	"runtime"
+	"unsafe"
+)
+
+type ReusableBuilder struct {
+	JSONBuilder
+}
+
+func (r *ReusableBuilder) Reset() *JSONBuilder {
+	r.JSONBuilder.Reset()
+	return &r.JSONBuilder
+}
+
+func (r *ReusableBuilder) UnsafeString() string {
+	b := r.Bytes()
+	return *(*string)(unsafe.Pointer(&b))
+}
+
+type JSONBuilder struct {
+	initialised bool
+	bytes.Buffer
+	gojson.Encoder
+}
+
+type noNewlineWriter struct {
+	*bytes.Buffer
+}
+
+func (w noNewlineWriter) Write(p []byte) (n int, err error) {
+	if len(p) > 0 && p[len(p)-1] == '\n' {
+		p = p[:len(p)-1]
+	}
+	return w.Buffer.Write(p)
+}
+
+// noescape hides a pointer from escape analysis. It is the identity function
+// but escape analysis doesn't think the output depends on the input.
+// noescape is inlined and currently compiles down to zero instructions.
+// USE CAREFULLY!
+// This was copied from the runtime; see issues 23382 and 7921.
+//
+//go:nosplit
+//go:nocheckptr
```

Comment:
really not a fan of point-in-time copies of runtime code that could become unsafe in the future